enclave/sgx/openssl_init.c

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/bits/types.h>
+
+/* Refer to openssl/crypto.h for the original definitions */
+#define OE_OPENSSL_INIT_NO_LOAD_CONFIG 0x00000080L
+
+/* Forward declarion */
+typedef struct oe_openssl_init_settings OE_OPENSSL_INIT_SETTINGS;
+int OE_OPENSSL_init_crypto(
+    uint64_t opts,
+    const OE_OPENSSL_INIT_SETTINGS* settings);
+
+/*
+ * Weak implementation that prevents link-time errors when an enclave
+ * does not link against OpenSSL libraries.
+ */
+int OE_OPENSSL_init_crypto(
+    uint64_t opts,
+    const OE_OPENSSL_INIT_SETTINGS* settings)
+{
+    OE_UNUSED(opts);
+    OE_UNUSED(settings);
+    return 0;
+}
+OE_WEAK_ALIAS(OE_OPENSSL_init_crypto, OPENSSL_init_crypto);
+
+/*
+ * Define the function as a constructor that ensures it is invoked before
+ * any OpenSSL APIs. Note that the constructor is only included by the linker
+ * when an enclave links against OpenSSL libraries (i.e., the
+ * OPENSSL_init_crypto symbol is included in the binary); i.e., when the enclave
+ * does not link against OpenSSL libraries, the constructor is discarded by the
+ * linker.
+ */
+__attribute__((constructor)) void oe_openssl_init()
```

Comment:
I don't think a constructor is the best solution here:
- Since no other compilation unit is calling any function defined in this file, the linker will not bring in any symbols from this file, including the constructor.
- Even if we force a dependency to this file, thereby retaining the constructor, the order of constructors is not predictable. We could work around it by explicitly specifying a priority for the constructor. But it just seems cleaner for oecore to call  `oe_crypto_init` (which could be a stub) at an appropriate point in initialization.

Reply:
For the current requirement, we don't need this constructor to be always linked; i.e., we only need the constructor when an enclave links against OpenSSL libraries. In this case, linker will always pick the constructor so we can ensure it runs before user's invocation.

We currently do need to enforce the execution order of this constructor and just need it to be present only an enclave links against OpenSSL. It does not depend on any constructor unless developer implements a constructor that uses OpenSSL APIs.

`oe_openssl_init` really indicates that it's specifically for OpenSSL, not Mbed. `oe_crypto_init` would be a bit confused given that we have `liboecrypto` already. However, the initialization for OpenSSL is needed even if the enclave is not using OpenSSL-based `liboecrypto`.

Reply:
>  we only need the constructor when an enclave links against OpenSSL libraries. In this case, linker will always pick the constructor so we can ensure it runs before user's invocation.

What causes the linker to pick up the constructor when linking against OpenSSL?

Reply:
OpenSSL also does lazy initialization by calling `OPENSSL_init_crypto`/`OPENSSL_ssl_crypto`. Therefore, these symbols are always present even if an enclave linking against OpenSSL, uses any OpenSSL APIs, but not calls them explicitly.